#162170668 User Signup Feature

marcdomain opened the pull request:
#### What does this PR do?
Handles implementation of user signup

#### Description of Task to be completed?
`POST /signup`
Use React and Redux on UI to signup users
Test the feature using `jest` and `react-testing-library`

#### How should this be manually tested?
- Clone this repository `git@github.com:andela/balder-ah-frontend.git`
- cd into the directory on your local machine
- On your command line, run `npm install` to install the project npm packages
- Run `npm run start:dev` to start the app
- On your web browser, visit the signup page on `/signup` and checkout the implementation with different inputs.
- Run `npm test` to test the app
- Run `npm test -- --coverage` to view the test coverage

#### Any background context you want to provide?
N/A

#### What are the relevant pivotal tracker stories?
[#162170668](https://www.pivotaltracker.com/story/show/162170668)

#### Screenshots
<img width="1430" alt="screen shot 2019-01-14 at 21 05 21" src="https://user-images.githubusercontent.com/25563661/51142271-c0eecc80-184b-11e9-9771-760f45c1e495.png">
<img width="1419" alt="screen shot 2019-01-14 at 21 06 51" src="https://user-images.githubusercontent.com/25563661/51142300-d4019c80-184b-11e9-8ac7-73ee8b5271a3.png">

<img width="1415" alt="screen shot 2019-01-14 at 21 07 42" src="https://user-images.githubusercontent.com/25563661/51142962-99006880-184d-11e9-8f1b-6ba9217782d9.png">
<img width="1419" alt="screen shot 2019-01-14 at 21 08 11" src="https://user-images.githubusercontent.com/25563661/51142967-9c93ef80-184d-11e9-919c-b86177fc87f2.png">

luqmanoop committed:
[Feature #162170671] User can login to Author's Haven

- Consume backend API to provide login for users
- Write test to cover for added components, actions and reducers

[Delivers #162170671]

omokaro-faith commented:
@marcdomain First things first Squash commits.

omokaro-faith commented on .env.sample:
```diff
@@ -0,0 +1 @@
+REACT_APP_API_URL=http://localhost:3000/api
```
Newline needed here.

omokaro-faith commented:
@marcdomain, Did you pull @codeshifu Branch in to work with?

omokaro-faith commented on src/actions/types.js:
```diff
@@ -0,0 +1,3 @@
+export default {
+  LOG_IN: 'LOG_IN',
+};
```
@marcdomain @codeshifu, I'll want all types in one file. We will be adapting Marcus's

omokaro-faith commented on src/reducers/signupReducers.js:
```diff
@@ -0,0 +1,31 @@
+    default:
+      return state;
+  }
+};
```
@marcdomain Thank you for not mutating state 💯

omokaro-faith commented on .gitignore:
```diff
@@ -21,3 +21,6 @@
 yarn-error.log*
+
+# environment variables
+.env
```
New line here also

omokaro-faith commented on src/App.jsx:
```diff
@@ -1,14 +1,15 @@
 import Home from './components/Home';
-import Login from './components/Login';
+import Login from './components/auth/login/Login';
+import Signup from './components/auth/signup/Signup';
```
Why do we have a signup folder ?

omokaro-faith commented on src/actions/signupActions/signup.js:
```diff
@@ -0,0 +1,22 @@
+    authUtils.saveUserToken(data.token);
+
+    dispatch({ type: SIGN_UP, payload: data });
+    dispatch({ type: SUCCESS_SIGNUP_MESSAGE, payload: data.message });
```
Why not destructure message and token.

omokaro-faith commented on src/components/auth/signup/Signup.css:
```diff
@@ -0,0 +1,104 @@
+  .formArea {
+    width: 100vw
+  }
+}
```
Newline here. Scss, please.

omokaro-faith commented on src/components/auth/signup/DisplayMessage.jsx:
```diff
@@ -0,0 +1,19 @@
+  </div>
+);
+
+export default DisplayMessage;
```
Nice job on creating a component for this.

omokaro-faith replied:
Since there is a presentation component folder already. You should include this to it instead of creating a new folder.

oluwajuwon replied:
yeah @marcdomain nice job here 👍

frostyblok commented on src/components/auth/signup/DisplayMessage.jsx:
```diff
@@ -0,0 +1,19 @@
+import React from 'react';
+
+const DisplayMessage = props => props.message && (
```
Hello Marcus, nice job here, but you could destructure 'message' and 'onSubmit' from the props, or what do you think?

Johnsonojo reviewed (changes requested):
@marcdomain, I noticed you're 2 commits on your PR. Please kindly fix it.

okoroemeka commented on src/components/auth/login/Login.css:
```diff
@@ -0,0 +1,75 @@
+.input-checkbox__label {
+  flex: 1;
+  margin-left: 5px;
+  /* START https://stackoverflow.com/a/7018415 */
```
what do these lines of code below do?

okoroemeka commented on src/components/auth/signup/Signup.jsx:
```diff
@@ -0,0 +1,146 @@
+
+  onChange = event => this.setState({ [event.target.name]: event.target.value });
+
+  comparePassword = () => {
```
Nice implementation @marcdomain . Love the fact that you are checking if the passwords match.

okoroemeka commented on src/components/auth/signup/Signup.jsx:
```diff
@@ -0,0 +1,146 @@
+    message: false,
+  }
+
+  onChange = event => this.setState({ [event.target.name]: event.target.value });
```
Would it not be nice if you destructure both name and value from `event.target`?. What do you think

akhilome commented on src/components/auth/signup/Signup.css:
```diff
@@ -0,0 +1,104 @@
+  background-size: cover;
+}
+
+.leftDiv {
```
Your use of `camelCasing` for CSS selectors is contradictory, style-wise, compared to how @codeshifu has already started off.

akhilome commented on src/components/auth/signup/Signup.jsx:
```diff
@@ -0,0 +1,146 @@
+
+        <div className="formArea">
+
+          <form className="form" onSubmit={this.onSubmit}>
```
Hmm. 🤔 

A `form`, having a class name of ... `form`?

akhilome reviewed (commented):
Hey @marcdomain, good job delivering on this task! 💪 

I voiced a few concerns below, please check them out.

marcdomain committed:
[Feat #162170668] user signup

  - Write tests for this feature
  - Create components for signup
  - consume signup API
